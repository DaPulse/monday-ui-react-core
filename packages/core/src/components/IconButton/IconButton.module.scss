.wrapper {
  width: fit-content;
  height: fit-content;
  display: inline-flex;
}

.referenceWrapper {
  display: inline-flex;
}

.loader.loader {
  width: 16px;
  height: 16px;
  &.xxs {
    width: 10px;
    height: 10px;
    svg {
      display: block;
    }
  }
}
