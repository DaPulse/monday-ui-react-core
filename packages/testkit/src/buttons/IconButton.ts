import { test, Page, Locator } from "@playwright/test";
import { Button } from "./Button";
import { Dialog } from "../popover/Dialog";

/**
 * Class representing an icon button that extends the Button class.
 */
export class IconButton extends Button {
  override page: Page;
  override locator: Locator;
  override elementReportName: string;

  /**
   * Create an IconButton.
   * @param {Page} page - The Playwright page object.
   * @param {Locator} locator - The locator for the IconButton element.
   * @param {string} elementReportName - The name for reporting purposes.
   */
  constructor(page: Page, locator: Locator, elementReportName: string) {
    super(page, locator, elementReportName);
    this.page = page;
    this.locator = locator;
    this.elementReportName = elementReportName;
  }

  /**
   * Select an item from the icon button's menu.
   * @param {string} item - The item to select.
   * @returns {Promise<void>}
   */
  async selectItem(item: string): Promise<void> {
    await test.step(
      `Select ${item} from ${this.elementReportName}`,
      async () => {
        await this.locator.click();
        const dialog = new Dialog(this.page, this.page.getByRole("dialog"), `${this.elementReportName} - Menu`);
        await dialog.selectItem(item);
      },
      { box: false }
    );
  }
}
