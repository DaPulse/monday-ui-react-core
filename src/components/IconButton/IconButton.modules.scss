.container {
  width: fit-content;
  height: fit-content;
}

.referenceContainer {
  display: inline-flex;
}