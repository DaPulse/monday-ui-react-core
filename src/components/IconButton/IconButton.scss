@import "../../styles/themes.scss";
@import "../../styles/typography.scss";

.icon-button-padding {
  margin-right: 8px;
  margin-left: 8px;
}
