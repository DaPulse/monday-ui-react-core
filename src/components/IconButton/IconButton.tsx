import { BUTTON_ICON_SIZE } from "../Button/ButtonConstants";
import React, { forwardRef, Fragment, useMemo, useRef } from "react";
import cx from "classnames";
import NOOP from "lodash/noop";
import useMergeRefs from "../../hooks/useMergeRefs";
import ToolTip from "../Tooltip/Tooltip";
import Icon from "../Icon/Icon";
import AddSmall from "../Icon/Icons/components/AddSmall";
import { getWidthHeight, Size } from "./services/IconButton-helpers";
import { getTestId } from "../../tests/test-ids-utils";
import { ComponentDefaultTestId } from "../../tests/constants";
import { VibeComponentProps, SubIcon } from "../../types";
import Button from "../Button/Button";
import { ButtonColor, ButtonType } from "../Button/ButtonConstants";
import styles from "./IconButton.module.scss";

export interface IconButtonProps extends VibeComponentProps {
  /**
   * id to be added to the element
   */
  id?: string;
  /**
   * callback function when clicking the icon button
   */
  onClick?: (event: React.MouseEvent) => void;
  /**
   * class to be added to the button
   */
  className?: string;
  /**
   * class to be added to the button wrapper
   */
  wrapperClassName?: string;
  /**
   * Icon to be rendered
   */
  icon?: SubIcon;
  /**
   * a11y property to be added, used for screen reader to know what kind of button it is
   */
  ariaLabel?: string;
  /**
   * Size of the icon
   */
  size?: Size;

  /**
   * Whether the tooltip should be displayed or not
   */
  hideTooltip?: boolean;
  /**
   * Tooltip wraps the button icon, it will display in the tooltip, if not present the aria label will be shown
   */
  tooltipContent?: string;
  /**
   * Kind of button - like <Button />
   */
  kind?: ButtonType;
  active?: boolean;
  /** The button's color  */
  color?: ButtonColor;
  /**
   * disabled state
   */
  disabled?: boolean;
  /**
   * if disabled - this will be shown in the tooltip
   */
  disabledReason?: string;
  dataTestId?: string;
  /** Change the focus indicator from around the button to within it */
  insetFocus?: boolean;
}

const IconButton: React.FC<IconButtonProps> & {
  sizes?: typeof Button.sizes;
  kinds?: typeof Button.kinds;
  colors?: typeof Button.colors;
} = forwardRef(
  (
    {
      className,
      wrapperClassName,
      id,
      icon,
      size,
      tooltipContent,
      ariaLabel,
      hideTooltip,
      kind,
      active,
      disabled,
      disabledReason,
      onClick,
      color,
      dataTestId,
      insetFocus
    },
    ref
  ) => {
    const componentRef = useRef(null);
    const mergedRef = useMergeRefs({ refs: [ref, componentRef] });
    const buttonAriaLabel = useMemo(() => {
      if (ariaLabel) return ariaLabel;
      if (typeof tooltipContent === "string") return tooltipContent;
      return undefined;
    }, [ariaLabel, tooltipContent]);

    const iconSize = useMemo(() => {
      switch (size) {
        case Button.sizes.XXS:
        case Button.sizes.XS:
          return 16;
        case Button.sizes.SMALL:
        case Button.sizes.MEDIUM:
        case Button.sizes.LARGE:
          return BUTTON_ICON_SIZE;
        default:
          return 24;
      }
    }, [size]);

    const overrideStyle = useMemo(() => {
      let style = {
        justifyContent: "center",
        alignItems: "center",
        padding: 0
      } as React.CSSProperties;
      if (active && kind !== IconButton.kinds.PRIMARY) {
        style.background = "var(--primary-selected-color)";
      }

      if (active && kind === IconButton.kinds.SECONDARY) {
        style.borderColor = "var(--primary-color)";
      }

      if (size) {
        style = { ...style, ...getWidthHeight(size) };
      }
      return style;
    }, [active, kind, size]);

    const calculatedTooltipContent = useMemo(() => {
      if (hideTooltip) return null;
      if (disabled && disabledReason) return disabledReason;
      if (tooltipContent) return tooltipContent;
      return ariaLabel;
    }, [hideTooltip, disabled, disabledReason, tooltipContent, ariaLabel]);

    const IconButtonWrapper = wrapperClassName ? "div" : Fragment;
    const iconButtonWrapperProps = useMemo(() => {
      return wrapperClassName ? { className: cx(wrapperClassName, styles.wrapper) } : {};
    }, [wrapperClassName]);

    return (
      <IconButtonWrapper {...iconButtonWrapperProps}>
        <ToolTip content={calculatedTooltipContent} referenceWrapperClassName={styles.referenceWrapper}>
          <Button
            onClick={onClick}
            disabled={disabled}
            color={color}
            kind={kind}
            ariaLabel={buttonAriaLabel}
            ref={mergedRef}
            id={id}
            dataTestId={dataTestId || getTestId(ComponentDefaultTestId.ICON_BUTTON, id)}
            noSidePadding
            active={active}
            className={className}
            style={overrideStyle}
            insetFocus={insetFocus}
          >
            <Icon
              icon={icon}
              iconType={Icon.type.SVG}
              iconSize={iconSize}
              ignoreFocusStyle
              className="icon-button-padding"
              clickable={false}
            />
          </Button>
        </ToolTip>
      </IconButtonWrapper>
    );
  }
);

Object.assign(IconButton, {
  sizes: Button.sizes,
  kinds: Button.kinds,
  colors: Button.colors,
  defaultTestId: ComponentDefaultTestId.ICON_BUTTON
});

IconButton.defaultProps = {
  className: undefined,
  wrapperClassName: undefined,
  onClick: NOOP,
  id: undefined,
  icon: AddSmall,
  ariaLabel: undefined,
  size: IconButton?.sizes.MEDIUM,
  hideTooltip: false,
  tooltipContent: undefined,
  kind: IconButton.kinds.TERTIARY,
  disabled: false,
  disabledReason: undefined,
  color: undefined,
  dataTestId: undefined,
  insetFocus: false
};

export default IconButton;
