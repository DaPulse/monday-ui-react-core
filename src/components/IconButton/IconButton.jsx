import React, { useRef, forwardRef, useMemo } from "react";
import cx from "classnames";
import PropTypes from "prop-types";
import useMergeRefs from "../../hooks/useMergeRefs";
import ToolTip from "../Tooltip/Tooltip";
import Button from "../Button/Button";
import Icon from "../Icon/Icon";
import "./IconButton.scss";

const IconButton = forwardRef(({ className, id, icon, size, tooltipContent, ariaLabel }, ref) => {
  const componentRef = useRef(null);
  const mergedRef = useMergeRefs({ refs: [ref, componentRef] });
  const buttonAriaLabel = useMemo(() => {
    if (ariaLabel) return ariaLabel;
    if (typeof tooltipContent === "string") return tooltipContent;
    return "";
  }, [ariaLabel, tooltipContent]);

  const iconSize = useMemo(() => {
    if (size === IconButton.sizes.SMALL) return 16;
    if (size === IconButton.sizes.MEDIUM) return 24;
    if (size === IconButton.sizes.LARGE) return 32;
    return 24;
  }, [size]);

  return (
    <ToolTip content={tooltipContent}>
      <Button
        size={size}
        kind={Button.kinds.TERTIARY}
        ariaLabel={buttonAriaLabel}
        ref={mergedRef}
        id={id}
        className={cx(className)}
        noSidePadding
      >
        <Icon
          icon={icon}
          iconType={Icon.type.SVG}
          iconSize={iconSize}
          ignoreFocusStyle
          className="icon-button-padding"
        />
      </Button>
    </ToolTip>
  );
});

IconButton.sizes = Button.sizes;
IconButton.propTypes = {
  id: PropTypes.string,
  className: PropTypes.string,
  icon: PropTypes.string,
  size: PropTypes.string,
  tooltipContent: PropTypes.string
};

IconButton.defaultProps = {
  className: "",
  id: ""
};

export default IconButton;
