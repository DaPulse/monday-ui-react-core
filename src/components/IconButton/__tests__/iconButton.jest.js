import React from "react";
import renderer from "react-test-renderer";
import IconButton from "../IconButton";

it("renders correctly with empty props", () => {
  const tree = renderer.create(<IconButton />).toJSON();
  expect(tree).toMatchSnapshot();
});
