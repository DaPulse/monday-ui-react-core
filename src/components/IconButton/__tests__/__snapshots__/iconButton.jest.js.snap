// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with empty props 1`] = `
<button
  aria-label=""
  className="monday-style-button monday-style-button--size-medium monday-style-button--kind-tertiary monday-style-button--color-primary monday-style-button--no-side-padding"
  disabled={false}
  id=""
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  type="button"
/>
`;
